x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationClusterService.java

```diff
@@ -0,0 +1,377 @@
+                            ),
+                            e
+                        );
+                        throw new ElasticsearchException("Error when starting node [{}] allocation", e, node.getId());
+                    }
+                }
+            }
+        }
+        return update(currentState, builder);
+    }
+
+    static ClusterState updateModelRoutingTable(ClusterState currentState, UpdateTrainedModelAllocationStateAction.Request request) {
+        final String modelId = request.getModelId();
+        final String nodeId = request.getNodeId();
+        TrainedModelAllocationMetadata metadata = TrainedModelAllocationMetadata.metadata(currentState);
+        logger.trace(
+            () -> new ParameterizedMessage("[{}] [{}] current metadata before update {}", modelId, nodeId, Strings.toString(metadata))
+        );
+        final TrainedModelAllocation existingAllocation = metadata.getModelAllocation(modelId);
+
+        // If state is stopped, this indicates the node process is closed, remove the node from the allocation
+        if (request.getRoutingState().getState().equals(RoutingState.STOPPED)) {
+            if (existingAllocation == null || existingAllocation.isRoutedToNode(nodeId) == false) {
+                return currentState;
+            }
+            return update(currentState, TrainedModelAllocationMetadata.builder(currentState).removeNode(modelId, nodeId));
+        }
+
+        if (existingAllocation == null) {
+            throw new ResourceNotFoundException("allocation for model with id [" + modelId + "] not found");
+        }
+        if (existingAllocation.isRoutedToNode(nodeId) == false) {
+            throw new ResourceNotFoundException("allocation for model with id [" + modelId + "] is not routed to node [" + nodeId + "]");
+        }
+        TrainedModelAllocationMetadata.Builder builder = TrainedModelAllocationMetadata.builder(currentState);
+        builder.updateAllocation(modelId, nodeId, request.getRoutingState());
+        return update(currentState, builder);
+    }
+
+    static ClusterState removeAllocation(ClusterState currentState, String modelId) {
+        TrainedModelAllocationMetadata.Builder builder = TrainedModelAllocationMetadata.builder(currentState);
+        if (builder.hasModel(modelId) == false) {
+            throw new ResourceNotFoundException("allocation for model with id [" + modelId + "] not found");
+        }
+        return update(currentState, builder.removeAllocation(modelId));
+    }
+
+    ClusterState addRemoveAllocationNodes(ClusterState currentState) {
+        TrainedModelAllocationMetadata previousState = TrainedModelAllocationMetadata.metadata(currentState);
+        TrainedModelAllocationMetadata.Builder builder = TrainedModelAllocationMetadata.builder(currentState);
+        Map<String, List<String>> removedNodeModelLookUp = new HashMap<>();
+        // TODO: make more efficient, right now this is O(nm) where n = sizeof(models) and m = sizeof(nodes)
```

Comment:
Another idea here would be to pass in the nodes diff instead of passing the whole cluster state. Then we could explicitly handle added and removed nodes.

Reply:
> Another idea here would be to pass in the nodes diff instead of passing the whole cluster state. Then we could explicitly handle added and removed nodes

True, but the view of the state from the `event` vs what is seen in the update call could be different. Its way safer to rely on the view of the state within the update callback vs passing in some external state